### PR TITLE
redo register assignment for SCfastpar and SCshadowreg parameters

### PR DESCRIPTION
**Pull request by WalterBright**
Turns out the way this was done before was too clever, it would prevent allocation of register parameters to the registers they came in. This fix should result in better register allocation on 64 bit targets. It doesn't matter for 32 bit targets as parameters are passed on the stack, not registers.

### CODE REVIEW

File: test/runnable/test_cdcmp.d [hunk 1]
```diff
@@ -1,7 +1,7 @@
 // REQUIRED_ARGS: -O -fPIC
 // PERMUTE_ARGS:
 // only testing on SYSV-ABI, but backend code is identical across platforms
-// DISABLED: win32 win64 osx linux32 freebsd32
+// DISABLED: win32 win64 osx linux32 freebsd32 linux64 freebsd64
```
Comment (ibuclaw) [inline test/runnable/test_cdcmp.d]: What is the point of this test if it's now disabled on all platforms dmd supports?

Comment (WalterBright) [reply test/runnable/test_cdcmp.d]: Because it's a highly annoying test.

1. It breaks every time I try to adjust the code generator.
2. Compiling the updating version of it is a nuisance.
3. Compiling the updating version requires the very latest Phobos to be installed and working.
4. Meanwhile, while trying to get (3) to work, the autotester won't test the rest of the compiler.
5. The compiler test suite shouldn't rely on Phobos, because that makes bootstrapping the compiler hard. Let alone the very latest Phobos (it requires std.typecons.EnumMembers).

I'll deal with it dead last.

Comment (wilzbach) [reply test/runnable/test_cdcmp.d]: > the very latest Phobos to be installed and working.

No that test is almost two years old, so the update script can't rely on "the very latest Phobos".

> The compiler test suite shouldn't rely on Phobos, because that makes bootstrapping the compiler hard

You do realize that the very script that you use to b execute the tests has been written in D? (d_do_tests)
Also, this test only requires Phobos for updating itself - not running.

**Review by wilzbach — DISMISSED** [review]
Let's not disable Martin's code gen tests without a good reason.

File: test/runnable/test_cdcmp.d (hunk 1 above)
Comment (wilzbach) [reply test/runnable/test_cdcmp.d]: DMD 2.060 (from 2012) has a bug fix for EnumMembers: 

https://dlang.org/changelog/2.060.html

I don't have time to look more into the past and you probably still need a recent version of DMD/Phobos, because for example everything before 2.072 was compiled without PIC meaning that you can't run it on any modern Linux distribution. Also our bootstrapping version of DMD is 2.079. That's what all CIs use too. We're not going to cater for anything below 2.079.

**Comment by WalterBright** [comment]
The "good reason" is because of a circular dependency on the compiler being tested to generate the tests, i.e. Phobos and the test updater have to be compiled with the compiler that is under test.

**Comment by WalterBright** [comment]
I'm planning some more alterations to the code generator after this one, which will just mess this up again. So I suggest deferring re-enabling these tests until the next release.

**Comment by wilzbach** [comment]
> The "good reason" is because of a circular dependency on the compiler being tested to generate the tests, i.e. Phobos and the test updater have to be compiled with the compiler that is under test.

No, that's absolutely not true. You can use any moderately recent Phobos version (like e.g. the one bundled with the Host DMD that you have to use to compile DMD).

**Comment by jacob-carlborg** [comment]
> The test should remain disabled until it is fixed to not require Phobos.

Or it should be fixed before adding more code that depends on it instead of disabling it.

File: test/runnable/test_cdcmp.d (hunk 1 above)
Comment (WalterBright) [reply test/runnable/test_cdcmp.d]: Basically all of Phobos has to work perfectly for this test to even compile, let alone run. It's just completely wrong for a back end test.

**Comment by WalterBright** [comment]
Or simply disable the test.

File: test/runnable/test_cdcmp.d (hunk 1 above)
Comment (andralex) [reply test/runnable/test_cdcmp.d]: @wilzbach that dependency seems onerous.

**Comment by andralex** [comment]
I'm okay with disabling the test. This way of validating compiler output is an ongoing job in and of itself.

**Comment by wilzbach** [comment]
> I'm okay with disabling the test. This way of validating compiler output is an ongoing job in and of itself.

Sorry, this doesn't make any sense to me. The test in question has been added by Martin about two years ago. We have never had any problems with it so far and it's the only way we have to test this. 
Disabling this is like saying we can't use D to run the testsuite, which we heavily use already
We can move the code to run these tests into d_do_test though which would also help to reduce the duplication.

Anyhow, as mentioned if we disable something that's been working well for years without any replacement, a better reasoning than principles is needed.

**Comment by WalterBright** [comment]
1. I've had trouble with it before, and updated it manually.
2. Again, it does not add value, it adds nuisance.
3. I haven't done much in the way of changing how the code generator works in the last couple years, except when I did resulting in (1).
4. I'm not deleting the test. Anyone can re-enable it in the future.

File: test/runnable/test_cdcmp.d (hunk 1 above)
Comment (wilzbach) [reply test/runnable/test_cdcmp.d]: > Basically all of Phobos has to work perfectly for this test to even compile, let alone run. 

__NOPE__: it requires no Phobos at all. Note the `version(update)`:

https://github.com/dlang/dmd/blob/ecbb547a826b4c198de78a91f002d94819d72b9d/test/runnable/test_cdcmp.d#L26-L28

It's only required for the automated update script.

**Comment by wilzbach** [comment]
> I've had trouble with it before, and updated it manually.

It comes with a handy auto-update feature. What's your issue with it?

> Again, it does not add value, it adds nuisance.

I strongly disagree, see e.g. https://github.com/dlang/dmd/pull/7003 or https://github.com/dlang/dmd/pull/7849 for the motivation behind these tests.

> I'm not deleting the test. Anyone can re-enable it in the future.

It's the same as deleting.

**Comment by WalterBright** [comment]
> see e.g. #7003 or #7849 for the motivation behind these tests.

They're poor motivation. Even Martin wrote: "I've disabled the OSX and Windows asm comparison tests completely. They use the same backend code, but due to ABI differences the assembly is different. Seems more pragmatic to just disable those tests instead of inventing some fuzzy matching."

**Comment by wilzbach** [comment]
> I've had trouble with it before, and updated it manually.

To clarify: I meant what's wrong with running the update script?

```
generated/linux/release/64/dmd -fPIC -version=update -run test/runnable/test_cdstrpar.d
```

It's easy and I just did that for you...
As you can see most tests gained one or two extra instructions, so I don't think that they are worthless.

If objdump is easier for you, I guess we can also convert these tests.

**Comment by wilzbach** [comment]
> They're poor motivation. Even Martin wrote: "I've disabled the OSX and Windows asm comparison tests completely. They use the same backend code, [...]

@MartinNowak's point was that they use the _same backend code_ which means that his added tests covered all platforms even if enabled only and thus avoided the need for fuzzy matching.

**Comment by WalterBright** [comment]
> I meant what's wrong with running the update script?

Please re-read what I wrote about running it.

**Comment by WalterBright** [comment]
> most tests gained one or two extra instructions

I know. Failing to pull this makes it difficult to work on the next part.

**Comment by WalterBright** [comment]
If you want those tests included, you're welcome to fiddle around with them after I'm done working on the register allocator. Not in the middle.

**Comment by wilzbach** [comment]
> If you want those tests included, you're welcome to fiddle around with them after I'm done working on the register allocator. Not in the middle.

The entire purpose of these tests is to avoid regressions which you're about to introduce.
Also, the master-stable branch off is about to happen today. So we would have pulled this codegen regression into the next release.

**Comment by WalterBright** [comment]
I also suggest you rewrite it so it does not require a perfect Phobos build.

**Comment by wilzbach** [comment]
> I also suggest you rewrite it so it does not require a perfect Phobos build.

Uhm. So the import issue was https://github.com/dlang/dmd/pull/9651, but I won't be able to fix this:

> compiling with version=update fails because because of the change in function signatures due to the dip1000 changes.

Reasons:
- if your host compiler Phobos library is broken, you won't be able to run the test suite anyhow (check the test script: https://github.com/dlang/dmd/blob/master/test/tools/d_do_test.d)
- `d_do_test.d` will actually (like the update script) be compiled with `HOST_DMD` which comes with a stable and working Phobos library
- it's super easy to get a working Phobos version on a Linux machine: 

```
curl https://dlang.org/install.sh
```

The only things that I could do are integration the update script with `d_do_test`, s.t. it doesn't get outdated and its works more seamless with `AUTO_UPDATE=1`.

File: test/runnable/test_cdcmp.d (hunk 1 above)
Comment (WalterBright) [reply test/runnable/test_cdcmp.d]: The update version has to be run with the compiler under test, and it relies on Phobos.

**Comment by WalterBright** [comment]
> curl https://dlang.org/install.sh

I never like those, because they never say what they're going to do:

1. where is it being installed?
2. what happens to any existing install?
3. that command isn't listed anywhere that I can find on dlang.org

I generally have multiple dmd release versions laying around. Running an 'install' program is usually incompatible with that.

I always hated installers anyway, going back to MSDOS where installers would usually trash your system. I'd fix Symantec C++ so you could run it off of the CD. No install required. It was the only compiler that didn't need installing. Every time I looked the other way, however, it would revert to needing to be "installed" before you could run it. Sigh.

File: test/runnable/test_cdcmp.d (hunk 1 above)
Comment (wilzbach) [reply test/runnable/test_cdcmp.d]: Nope, you set the `DMD` under test via an environment (as the update script header explains) and can thus continue to use the host compiler for compiling the update script.

**Comment by WalterBright** [comment]
Anyhow, I often use a script to run the test suite, because it doesn't require Phobos. If you've ever tried to bootstrap a compiler on a new system, the value of this will be clear.

**Comment by wilzbach** [comment]
> that command isn't listed anywhere that I can find on dlang.org

It's the first hit on Google:

image

> I generally have multiple dmd release versions laying around. Running an 'install' program is usually incompatible with that.

I highly recommend you to read the documentation of the _official_ D install script. It's just here:

https://dlang.org/install

tl;dr: it is not incompatible.
- by default it will "install" (aka download) compilers to `~/dlang`
- you can use them directly with e.g. `~/dlang/dmd-2.085.0/linux/bin64/dmd`
- `activate` will overload your current PATH with the chosen DMD binary

Anyhow, it's beside the point as you need to have a host dmd to build DMD anyhow (and run the testsuite).

> If you've ever tried to bootstrap a compiler on a new system, the value of this will be clear.

Sure, but since DMD is now written in D one needs a host DMD anyhow when to bootstrap the compiler to a new platform. As the `dmd-cxx` branch is in a really bad shape, the current best practice approach is to use LDC to cross-compile the compiler to the new platform. If you can cross-compile the compiler, then you can also cross-compile Phobos which you need to do anyhow for port to work with any non-trivial D script.

> Anyhow, I often use a script to run the test suite, because it doesn't require Phobos.

Well, everyone else is using `test/run.d` (or the older `test/Makefile`). Please don't expect things to work otherwise.

Here's what everyone is doing:

1) Install host compiler (e.g. via package manager, install script, download binaries directly, ...)
2) Build DMD with the host compiler (by default `HOST_DMD=dmd` - i.e. the globally installed dmd instance)
3) Build Test suite runner with the host compiler

(this is simplified because you still need to build druntime and phobos to get a libphobs2.a / libphobos2.so that will be linked with all runnable test even if it's not used.)

`HOST_DMD` != the currently tested D compiler.

I absolutely understand that for debugging tests, it's very helpful not to depend on Phobos, but the build system and tools around the tests can use Phobos (as long as they are using stable parts of it, which we check with e.g. SemaphoreCI).

> The only things that I could do are integration the update script with d_do_test, s.t. it doesn't get outdated and its works more seamless with AUTO_UPDATE=1.

I did this in https://github.com/dlang/dmd/pull/9653 (this removes all Phobos imports from the codegen files).

**Comment by andralex** [comment]
@wilzbach you are making good points, and the decision would be obvious if we had a staff of compiler engineers. Then maintaining the tests would be naturally part of someone's ongoing duties. As such, we're essentially arguing how @WalterBright is to spend his time and what is and what isn't good, easy, or productive for him. This doesn't seem fair. 

The short of it is: If we pull this and then somebody carefully analyzes the code, makes sure it's all dandy, and then runs the updater and re-enables the tests - then that would be neat. If you're up for it please let it be known!

**Comment by wilzbach** [comment]
No, it shouldn't never ever be the job of someone else to update the tests which Walter broke. This will never work and just lead to a lot of frustration and broken tests. I understand that Walter wants this change in master, but it's a codegen regression which the tests have clearly shown which justifies their purpose more than enough already.

> As such, we're essentially arguing how @WalterBright is to spend his time and what is and what isn't good, easy, or productive for him.

Sure, but we even bother with CIs or reviews then? They just make things less productive and harder.
And if we're at it, why even listen to the community (not that this is actually happening frequently anyhow anymore)? It's less productive too.

> makes sure it's all dandy, and then runs the updater and re-enables the tests - then that would be neat. If you're up for it please let it be known!

Nope, not up for that _at all_. Running the update script takes less than five seconds and looking at it's changes is __vitally__ important for reviewing the changes here. I don't see how anyone could blindly approve codegen changes without at least roughly seeing what the changes result into.

Also, note that I actually did update the test scripts for Walter in this PR, but he just force-pushed over it :shrug:

**Comment by MartinNowak** [comment]
It's kind of a weird discussion that we already had when the tests were added.
https://github.com/dlang/dmd/pull/7849

The point Sebastian is making are all correct, the driver to update the script uses the host compiler and some phobos stuff for regexing, the latest compiler is only needed as the test subject to generate the code.
I can also understand that Walter is reluctant to pick up any tooling given the amount of time he is already investing, but it can lead to the false perception that it's due to ignorance. Also Walter is mostly working on the weirder side of D toolchains (Windows), so there is extra effort for him to update on Linux.
Dropping the tests is debatable. They are very helpful for developing (e.g. the SIMD initialization codegen), prevent accidental regressions, and help to review and document changes (e.g. https://github.com/dlang/dmd/commit/fed96c532304d2be1b6ee1e9825e6257c776302e#diff-e5a7224f0f623f184e3b6b13bb04149f).

Let's look at recent edits to those files.
- https://github.com/dlang/dmd/commits/master/test/runnable/test_cdvecfill.d
  No change after initial addition except for some PIC+non-PIC detour
- https://github.com/dlang/dmd/commits/master/test/runnable/test_cdcmp.d
  Three codegen improvements in 1.5 years.
- https://github.com/dlang/dmd/commits/master/test/runnable/test_cdstrpar.d
  One REX-prefix related change in 1.5 years plus the first attempt at sneaking in a disabling of the tests.
  (https://github.com/dlang/dmd/pull/7912#pullrequestreview-98841729)

Some ideas to avoid changing the tests on pre-/postamble codegen changes would be very welcome, definitely more constructive than *this* kind of poisonous discussions ;).
Maybe some `asm { naked; }` would already do the trick?

I also don't understand why this is not a straight discussion on a PR deleting proposing to delete those tests. That would have helped people to properly understand the usability problems and either accept deletion or resolve them.
Sneaking in test disabling with some PRs is not a practice anyone wants to establish, is it?
https://github.com/dlang/dmd/pull/7912#pullrequestreview-98841729
https://github.com/dlang/dmd/pull/9608/files#diff-6ae615a8fd4d30180e88171bfcffb97bR4

As I've said in the past I don't see that the backend provides much value differentiation for D.
For me it's Walter's personal pet peeve, so I stopped caring about it long ago. What I do see is that being perceived as part of D it consistenly sucks up lots of contributions, development time, and discussions.

**Comment by WalterBright** [comment]
The current crop of test failures is due to an unrelated bug this PR uncovered: https://github.com/dlang/dmd/pull/9656
which is blocking this PR.

**Comment by WalterBright** [comment]
> It's the first hit on Google:

It might as well not exist if it is not linked to from dlang.org.

**Comment by WalterBright** [comment]
> I stopped caring about it long ago

Then please allow me to develop it as I see fit.

**Comment by jacob-carlborg** [comment]
> It might as well not exist if it is not linked to from dlang.org.

It is: https://dlang.org/download.html

**Comment by WalterBright** [comment]
So the most important link is way beyond the fold, that looks like this:
```
Install Script (i)
curl -fsS https://dlang.org/install.sh | bash -s dmd-nightly
```
Need to give some exposition here. I recommend adding this text:
```
The dlang.org/install.sh script is the official D version manager for DMD, GDC and LDC on FreeBSD, Linux and macOS. It provides a convenient way to fetch and install a D compiler without requiring administrative privileges nor external dependencies. (i)
```
And leave off the bash command. The installer documentation makes no mention of `-fsS`, either. Why is that there?

**Comment by WalterBright** [comment]
I've added a bug report for the installer https://issues.dlang.org/show_bug.cgi?id=19815 so please continue any discussion of that there, not here.

**Comment by WalterBright** [comment]
I propose a simple solution:

Rewrite test_cdcmp.d to a simple sequence of:
```
bool test1(ubyte x) { return x < 0; }
bool test2(ubyte x, ubyte y) { return x < y; }
...
```
Write a script that:
1. compiles test_cdcmp.d with -c -O
2. runs objdump on test_cdcmp.o and writes the output to test_cdcmp.out
3. diffs test_cdcmp.out with the known good output, test_cdcmp.txt

That's it. It doesn't depend on the dmd toolchain working. Even better, the diff of the files gives you all the differences at once. "Updating" is as trivial as:
```
cp test_cdcmp.out test_cdcmp.txt
```

**Comment by WalterBright** [comment]
Proposed solution: https://github.com/dlang/dmd/pull/9668

**Comment by wilzbach** [comment]
> That's it. It doesn't depend on the dmd toolchain working. 

The current solution doesn't depend on the dmd toolchain either when _running_ in the testsuite. Furthermore, it doesn't depend on outside tools like `objdump` that very likely will have different output from different versions and output. After all, that's why your proposed PRs are failing.


Have you seen https://github.com/dlang/dmd/pull/9653? It makes the test script work with `AUTO_UPDATE` (a lot of the code could be dropped by iterating through all `testee__` strings in the output, but that refactoring PR tried to keep to the original solution as close as possible).

Furthermore,

1) As Mike already mentioned, we're actively working on getting rid of shell scripts as they don't work great on all platforms.

2) Martin's approach allows us to add matching for OSX, FreeBSD etc. because we're in control of writing the test output and comparing the output.

3) It won't work with the existing `AUTO_UPDATE`

4) and this is the most important point: the existing solution doe _not_ use `objdump`  when run in the test suite. This means it passes without any problems on any kind of Linux 64-bit system. This works because it uses plain D to inspect its own object code. As the failures on the PRs of your proposed solutions show, this is a very tricky problem.

> Then please allow me to develop it as I see fit.

The problem is that all the newcomers to D are affected by DMD as we still call it "reference compiler".
We should have switched to Clang a long time ago. D as a business would have, because it allows to use the small manpower available to focus on the important tasks at hand that the customers actually care about. D as an open source project is just moving in its own limbo / dying its slow death instead.

However, as long as DMD stays the recommended compiler on dlang.org, we will have to waste our time by e.g. this objection to DMD codegen regressions.

**Comment by WalterBright** [comment]
> After all, that's why your proposed PRs are failing.

It is not. It was failing because I didn't pass -m64 to dmd in the script. As you can see, they work now. They are trivial scripts - 4 commands. They are not a maintenance problem.

> As the failures on the PRs of your proposed solutions show, this is a very tricky problem.

It is not tricky at all. I forgot to pass -m64. I made a simple oversight - look at the failure: it is because the diff is between the output of -m32 and the output of -m64.

>  doe not use objdump when run in the test suite.

It relies on objdump to update the test program. I don't object to it using objdump - it's sensible to use it - just that *both* methods use objdump.

> we will have to waste our time

The true waste of time is replacing 4 lines of trivial script with 200+ lines of fairly complex D code that uses half of Phobos. The code in it is clever, I agree. But it is solving the problem by going around the horn. We should take a step back and not get enthralled by complex solutions when simple ones are available. I bet half of D programmers could not figure out how the complex solution works, and the rest would spend hours working it out. The 4 commands in the script are apparent even to people who don't know bash. (I don't know bash, and I wrote it in a few minutes.)

**Comment by MartinNowak** [comment]
> Write a script that:
> 
>     1. compiles test_cdcmp.d with -c -O
> 
>     2. runs objdump on test_cdcmp.o and writes the output to test_cdcmp.out
> 
>     3. diffs test_cdcmp.out with the known good output, test_cdcmp.txt

Sounds reasonable to me ;).

**Comment by WalterBright** [comment]
This change usually (but not always) produces better code, as the diffs show. The larger code generated examples are all due to the same effect - live register analysis is blocking using the fastpar register directly. I don't have a good solution to that at the moment, but this PR is still overall an improvement, and I don't want to make it even more complex.

**Comment by WalterBright** [comment]
@wilzbach @thewilsonator this is ready now.

**Review by andralex — APPROVED** [review]
I hope the controversy has been put to sleep now. Thanks!

**Comment by ibuclaw** [comment]
This PR introduced a regression https://issues.dlang.org/show_bug.cgi?id=23084
